README.md
nightmare-uber
=================

[Nightmare](https://github.com/segmentio/nightmare) plugin for [m.Uber.com](https://m.uber.com). Requires `node 0.11.x` and `node --harmony` to run. Please refer to the [examples](https://github.com/lambtron/nightmare-uber/tree/master/examples) for more information.

#### .login(email, password)

Login to Uber. Note that this must be called first.

#### .select(type)

_(Untested)_ Select the type of Uber: `uberpool`, `uberx`, `uberxl`, `blackcar`, `suv`, or `taxi`. Note the type is a case insensitive string. Note that this should be called before the other functions, as you aren't able to change this option after `.setPickup()` or `.setDropoff()` are called.

#### .setPickup(location)

_(Untested)_ Set the pick up location. The `location` can either be an object with `lat` and `lng` properties (each a float value) or a string of an address. If `lat` and `lng` are passed, they will be resolved to an address via reverse geocoding.

This must be called prior to `.setDropoff()`.

#### .setDropoff(location)

_(Untested)_ Set the drop off location with similar methods applied in `.setPickup()`.

#### .getFare(cb)

_(Untested)_ Get fare estimates. Assumes that both `setPickup()` and `setDropoff()` have been called. Callback expression is `cb(fare)`.

#### .call()

_(Untested)_ Call the car. Note that `.setPickup()` must have been called; if you selected `uberpool`, then `.setDropoff()` must also have been called.

## License (MIT)

```
WWWWWW||WWWWWW
 W W W||W W W
      ||
    ( OO )__________
     /  |           \
    /o o|    MIT     \
    \___/||_||__||_|| *
         || ||  || ||
        _||_|| _||_||
       (__|__|(__|__|
```

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



History.md

0.0.1 / 2014-12-01
==================

  * setting up package.json
  * added basic functions
  * updating readme
  * Initial commit


test/test.js
var Nightmare = require('nightmare');
var LinkedIn = require('../index.js');

var nightmare = new Nightmare();
var email = 'andyjiang@gmail.com';
var password = 'mengfei8*';

var a = './test/a.png';
var b = './test/b.png';
var c = './test/c.png';

nightmare
  .viewport(800, 1600)
  .goto('https://m.uber.com/')
  .wait('a.done')
  .screenshot(a)
  .wait(1000)
  .type('#email', email)
  .type('#password', password)
  .screenshot(b)
  .wait(1000)
  .click('a.done')
  .wait('.show-eta')
  .wait(10000)
  .screenshot(c)
  .run(function(err, nightmare) {
    console.log('done');
  });

test/index.js

/**
 * Module dependencies.
 */

var Nightmare = require('nightmare');
var Uber = require('../index.js');
var should = require('should');
var after = require('after');

/**
 * Test.
 */

describe('Nightmare-Uber', function() {
  this.timeout(100000);

});

index.js

/**
 * Dependencies.
 */

var exec = require('child_process').exec;
var thunkify = require('thunkify-wrap');
var Geocoder = require('geocoder');
var geo = thunkify(Geocoder);

/**
 * Login to an Uber account.
 *
 * @param {String} email
 * @param {String} password
 */

var login = exports.login = function(email, password){
  return function(nightmare) {
    nightmare
      .viewport(800, 1600)
      .goto('https://m.uber.com/')
      .wait('a.done')
      .wait(1000)
      .type('#email', email)
      .type('#password', password)
      .click('a.done')
      .wait('.show-eta')
      .wait(10000)
      .evaluate(function() {
        var overlay = document.querySelector('.dialog-overlay');
        var parent = overlay.parentNode;
        parent.removeChild(overlay);
      }, function() {
      });
  };
};

/**
 * Select type of car.
 *
 * @param {String} type
 */

var select = exports.select = function(type) {
  var mapping = {
    uberpool: 1,
    uberx: 2,
    uberxl: 3,
    blackcar: 4,
    suv: 5,
    taxi: 6
  };
  var nth = mapping[type.toLowerCase()];
  return function(nightmare) {
    nightmare
      .click('div.vehicle-selector > ul >li.nth-child(' + nth + ')')
      .wait(100);
  };
};

/**
 * Set pick up location (lat & lng).
 *
 * @param {String} location (address)
 * @param {Object} location (lat, lng)
 */

var setPickup = exports.setPickup = function *(location) {
  var address = yield getAddress(location);
  return function(nightmare) {
    nightmare
      .type('input[name="search"]', address)
      .wait(1000)
      .click('.search-results > ul:nth-child(1) > li > a')
      .wait('.set')
      .wait(500)
      .click('.set');
  };
};

/**
 * Set drop off location (lat & lng)
 *
 * @param {String} location (address)
 * @param {Object} location (lat, lng)
 */

var setDropoff = exports.setDropoff = function *(location) {
  var address = yield getAddress(location);
  return function(nightmare) {
    nightmare
      .exists('a.btn', function(ready) {
        if (!ready) throw 'Set drop off box not ready';
        return;
      })
      .click('a.btn')
      .type('input[name="search"]', address)
      .click('.search-results > ul:nth-child(1) > li > a')
      .wait('a.accept')
      .wait(500);
  };
};

/**
 * Get fare estimates.
 *
 * @param {Function} callback
 */

var getFare = exports.getFare = function(callback) {
  return function(nightmare) {
    nightmare
      .click('a.fare-quote')
      .evaluate(function() {
      }, function(fare) {
        callback(fare);
      });
  };
};

/**
 * Call Uber
 *
 * TODO: improve
 */

var call = exports.call = function() {
  return function(nightmare) {
    nightmare
      .exists('a.accept', function(ready) {
        if (!ready) throw 'Unable to call Uber.';
        return;
      })
      .click('a.accept');
  };
};

/**
 * Private function to reverse geocode lat, lng.
 *
 * @param {Integer} lat
 * @param {Integer} lng
 */

function *reverseGeocode(lat, lng) {
  var geos = yield geo.reverseGeocode(lat, lng);
  return geos.results[0].formatted_address;
}

/**
 * Private function to get address from location.
 *
 * @param {String or Object} location
 */

function *getAddress(location) {
  var address = location;
  if (location.lat && location.lng)
    address = yield reverseGeocode(location.lat, location.lng);
  return address;
}
